test/tf_exec.cpp

```diff
@@ -104,4 +103,82 @@ TEST(tf_exec, axpy) {
   }
 }
 
+void AssertTensorEquals(tf::Tensor T1, tf::Tensor T2) {
+  auto T_size = T1.flat<float>().size();
+  for (int k=0; k<T_size; k++) {
+    auto a = T1.flat<float>().data()[k];
+    auto b = T2.flat<float>().data()[k];
+    EXPECT_FLOAT_EQ(a, b);
+  } 
+}
+
+TEST(tf_exec, BatchMatMul) { 
```

Comment:
Let's add some more unit tests for some trickier cases, specifically ones where the lengths of the batch axes don't happen to match either of the matrix dimensions. (For example, try `{2,3,4,5}` with `{2,3,5,4}`. Probably will want to generate the coefficients in a loop, of course. :))

Reply:
Also, we should run some weird corner cases like ones where the batch or matrix sizes are zero. For example, both of these should work:

```
>>> from tensorflow.python.ops.gen_math_ops import batch_mat_mul

>>> x = tf.constant([],shape=[2,0,2,2])
>>> y = tf.constant([],shape=[2,0,2,2])
>>> tf.Session().run(batch_mat_mul(x,y))
array([], shape=(2, 0, 2, 2), dtype=float32)

>>> x = tf.constant([],shape=[2,3,2,0])
>>> y = tf.constant([],shape=[2,3,0,2])
>>> tf.Session().run(batch_mat_mul(x,y))
array([[[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]],


       [[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]]], dtype=float32)
```

Reply:
The new tests look good! However, I think it is probably better to split them up into separate tests for the individual cases.

Reply:
Split the test cases to 0D, 2D, 3D and general